### ISSUE TITLE
Refactor predict_proba

### ISSUE THREAD

**Issue by luizgh** [Menelau/DESlib#55]
Create integration tests with predict_proba
Run predict_proba on batches. 
On "predict", pass the probabilities to Meta-DES and KNOP

**Comment by luizgh** [comment]
@Menelau - I implemented most of the functionality on commit adfcd7c21f3f2fc6077eb00abf4887cd2d567880. There are still some tests failing on test_integration_dfp, that I did not have time to check.

By the way, I noticed that there were no integration tests using the "weighting" or "hybrid" approaches (only "selection"), so we still need to test those features. In particular, I needed to refactor the "predict_proba_ensemble_weighted" function (from util.aggregation) to work with masked arrays (I opted for an implementation without using einsum, but I need your help to verify it is implemented correctly).

**Comment by Menelau** [comment]
@luizgh Ok I will check that,

About the other integration tests, there are some using the "weighting" mode since the KNORA-U and KNOP are based on this mode (they only use the weighted version). For the hybrid yes we need to include a few tests also considering this mode.

For the "predict_proba_weighted" I can confirm that with some examples that I have from the textbook "Combining pattern classifiers". Did it pass the aggregation unit tests?

**Comment by luizgh** [comment]
Yes, it is passing all unit tests. There are no tests on test_aggregation for this function (predict_proba_ensemble_weighted), but I noticed that the function is being tested on the DES/test_base (test_predict_proba_weighting, test_predict_proba_hybrid)

**Comment by luizgh** [comment]
@Menelau I found one of the issues with the DFP tests with proba (I had forgot to apply the DFP mask, now that it is no longer done in the "estimate_competences" method.

I fixed that, but I am still seeing errors. For example, the test for DESP is failing ( test_desp_proba in the file test_integration_dfp.py). I noticed that for example 6, the estimation of the competences is different between the old and new implementations. The DFP mask is also different.

Old implementation:
competences: [ 0.85714286  0.85714286  0.85714286  1.          1.          1.          1.
  0.          0.85714286  0.85714286]
dfp mask: [ 1.  1.  1.  1.  1.  1.  1.  0.  1.  1.]

New implementation:
competences[1] (original index: 6): [ 0.85714286  0.85714286  0.85714286  0.85714286  0.85714286  0.85714286  1.          0.71428571  1.          1.        ]
dfp mask: [ 1.  1.  1.  1.  1.  1.  1.  1.  1.  1.]

**Comment by Menelau** [comment]
Are you sure you are seeing the same entries of the array? It may be confusing due to how the indices are managed in the new version. The DFP mask should be the same for all DS methods in this same problem (for the predict and predict_proba). I will check the new commits you made in the batch_proba branch for the predict proba method, however comparing the DFP mask for the DES-P method (old and new I'm receiving getting the same mask and competences array (attached file). 

[competences.zip](https://github.com/Menelau/DESlib/files/1885635/competences.zip)
[DFP mask.zip](https://github.com/Menelau/DESlib/files/1885609/DFP.mask.zip)

The mask and competences should be equals to these files independently of the method used (predict or predict_proba). We could also check if the problem was not with the old implementation of predict proba

**Comment by Menelau** [comment]
@luizgh Found the error, the expected values were wrong due to an error with the competence estimation in the previous version of the code that was used to generate the expected values!. Found that since both the matrix with estimated competence and the DFP mask was a bit different between the predict and predict_proba methods (which should always be exactly the same).  I've updated the expected values in the newest commit ecc7c99b67f122f276aaabf729a990bef70eaed6

Now it seems all tests with predict_proba are passing. Sorry about that, it was just a very stupid mistake!
